src/index.ts

```diff
@@ -1,47 +1,70 @@
-      if (property.format === "date-time") {
-        return "time.Time";
-      }
-      return "string";
-    case "number":
-      if (property.format === "float") {
-        return "float32";
-      }
-      if (property.format === "double") {
-        return "float64";
+function toGolangTypeX(type: XtpNormalizedType): string {
+  // turn into reference pointer if needed
+  const pointerify = (t: string) => {
+    return `${type.nullable ? '*' : ''}${t}`
+  }
+
+  switch (type.kind) {
+    case 'string':
+      return pointerify('string')
+    case 'int32':
+      return pointerify('int32')
+    case 'float':
+      return pointerify('float32')
+    case 'double':
+      return pointerify('float64')
+    case 'byte':
+      return pointerify('byte')
+    case 'date-time':
+      return pointerify('time.Time')
+    case 'boolean':
+      return pointerify('bool')
+    case 'array':
+      const arrayType = type as ArrayType
+      return pointerify(`[]${toGolangTypeX(arrayType.elementType)}`)
+    case 'buffer':
+      return pointerify('[]byte')
+    case 'object':
+      const oType = (type as ObjectType)
+      if (oType.properties?.length > 0) {
+        return pointerify(goName(oType.name))
+      } else {
+        // let's use empty interface for an untyped object
+        return pointerify("interface{}")
       }
-      return "int64";
-    case "integer":
-      return "int32";
-    case "boolean":
-      return "bool";
-    case "object":
-      return "map[string]interface{}";
-    case "array":
-      if (!property.items) return "[]any";
-      // TODO this is not quite right to force cast
-      return `[]${toGolangType(property.items as Property)}`;
-    case "buffer":
-      return "[]byte";
+    case 'enum':
+      return pointerify(goName((type as EnumType).name))
+    case 'map':
+      const { keyType, valueType } = type as MapType
+      return pointerify(`map[${toGolangTypeX(keyType)}]${toGolangTypeX(valueType)}`)
     default:
-      throw new Error("Can't convert property to Go type: " + property.type);
+      throw new Error("Can't convert XTP type to Go type: " + type)
   }
 }
 
-function pointerToGolangType(property: Property) {
-  const typ = toGolangType(property);
+function toGolangType(property: Property, required?: boolean): string {
+  const t = toGolangTypeX(property.xtpType)
 
-  if (typ.startsWith("[]") || typ.startsWith("map[")) {
-    return typ;
-  }
+  // if required is unset, just return what we get back
+  if (required === undefined) return t
+
+  // if it's set and true, just return what we get back
+  if (required) return t
+
+  // otherwise it's false, so let's ensure it's a pointer
+  if (t.startsWith('*')) return t
+  return `*${t}`
+}
+
+// used to define a return type, for backwards compat
+function toGolangReturnType(property: Property): string {
+  const t = toGolangTypeX(property.xtpType)
 
-  return `*${typ}`;
+  if (t.startsWith('[]') || t.startsWith('map[')) return t
+  return `*${t}`
 }
```

Comment:
This is a little bit awkward but it preserves the "ergonomic" return type logic we had previously. I think it's worth keeping as it's more idiomatic to go over simplicity in the code. though maybe there is a better name or way i can achieve simplicity here.